cpp_build/src/lib.rs

```diff
@@ -44,7 +44,47 @@ const INTERNAL_CPP_STRUCTS: &'static str = r#"
 
 #include "stdint.h" // For {u}intN_t
 #include <new> // For placement new
-
+#include <utility>
+
+namespace rustcpp {
+
+// We can't just pass or return any type from extern "C" rust functions (because the call
+// convention may differ between the C++ type, and the rust type).
+// So we make sure to pass trivial structure that only contains a pointer to the object we want to
+// pass. The constructor of these helper class contains a 'container' of the right size which will
+// be allocated on the stack. The use of an anonymous union allow to call manually the constructor
+// and destructor as needed.
```

Comment:
Can we just put the value we're capturing behind a raw pointer? I don't really like these templates here.

The only use case I see this really helping with is reference types, which I think we can ask people to use pointers for.